### PR TITLE
use a grin specific tmp dir instead of a system tmp dir

### PR DESCRIPTION
**Pull request by garyyu**
Fix for https://github.com/mimblewimble/grin/issues/2709

And improvement according to some recommendations:
- Randomize temporary file names
- Only write to files in directories owned and writable by the node user
- Using the ​`create_new`​ option from `​std::fs::OpenOptions`​ instead of `create`

TODO:
- [X] After https://github.com/mimblewimble/grin/pull/2685 merged, the `sandbox` dir also need switch to grin specific tmp dir.

### CODE REVIEW

File: p2p/src/protocol.rs [hunk 1]
```diff
@@ -288,10 +288,24 @@ impl MessageHandler for Protocol {
 				self.adapter
 					.txhashset_download_update(download_start_time, 0, sm_arch.bytes);
 
-				let mut tmp = env::temp_dir();
-				tmp.push("txhashset.zip");
+				let mut tmp = self.adapter.get_tmp_dir();
```
Comment (ignopeverell) [inline p2p/src/protocol.rs]: Perhaps now that logic would fit better in the chain, given that it already holds the preparation of the directory location?

File: servers/src/common/adapters.rs [hunk 2]
```diff
@@ -369,6 +370,19 @@ impl p2p::ChainAdapter for NetToChainAdapter {
 			true
 		}
 	}
+
+	/// Specific tmp dir.
+	/// Normally it's ~/.grin/main/tmp for mainnet
+	/// or ~/.grin/floo/tmp for floonet
+	fn get_tmp_dir(&self) -> PathBuf {
```
Comment (ignopeverell) [inline servers/src/common/adapters.rs]: Why not delegate to the chain implementation here? Am I missing something?

Comment (garyyu) [reply servers/src/common/adapters.rs]: accept

File: p2p/src/protocol.rs (hunk 1 above)
Comment (garyyu) [reply p2p/src/protocol.rs]: What do you mean?  moving `save_txhashset_to_file()` function into `chain`?

**Comment by yeastplume** [comment]
Something else needs to be verified here, when trying to test 1.1.0 I came across an issue on this line:

https://github.com/mimblewimble/grin/blob/325e32821dad60c0509ba938e7b0f5001aaa161c/chain/src/txhashset/txhashset.rs#L1466

The error I'm getting is:

`20190402 08:09:51.600 ERROR grin_chain::txhashset::txhashset - hashset_replace fail on txhashset. err: Invalid cross-device link (os error 18)`

What's actually happening is that the tmp dir is a seperate tmpfs mount, so rename won't work across mounted devices. This fix definitely needs to go into the 1.1.0 release.

File: p2p/src/protocol.rs (hunk 1 above)
Comment (garyyu) [reply p2p/src/protocol.rs]: @ignopeverell  remind ^ above question.

Comment (ignopeverell) [reply p2p/src/protocol.rs]: I meant that whole code section, with getting the tmp dir, checking if it exists, creating it if not, appending file name, etc.

Comment (garyyu) [reply p2p/src/protocol.rs]: done. moved the logic to chain

**Comment by garyyu** [comment]
@ignopeverell  ready for another review, thanks

**Comment by ignopeverell** [comment]
Given where we're at, how about targeting this for 1.1.0?

**Comment by garyyu** [comment]
ok, how to move an existing PR to another branch?

**Comment by garyyu** [comment]
Replaced by #2753 for branch 1.1.0

**Pull request closed without merging**